src/components/SigningModal/SigningModal.js

```diff
@@ -0,0 +1,66 @@
+import React, { useState, useEffect, useRef } from 'react'
+import PropTypes from 'prop-types'
+import { AppType } from '../../prop-types'
+import SignMessageFlow from './SignMessageFlow'
+import SignTransactionFlow from './SignTransactionFlow/SignTransactionFlow'
+
+function SigningModal({ apps, transactionBag, signatureBag, web3 }) {
+  const [modalVisible, setModalVisible] = useState(false)
+  const [isTransaction, setIsTransaction] = useState(false)
+
+  const prevTxBag = useRef()
+  const prevSigBag = useRef()
+
+  const sigBagUpdate =
+    signatureBag && signatureBag !== prevSigBag.current && signatureBag
+  const txBagUpdate =
+    transactionBag && transactionBag !== prevTxBag.current && transactionBag
+
+  useEffect(() => {
+    if (txBagUpdate) {
+      setModalVisible(true)
+      setIsTransaction(true)
+      prevTxBag.current = txBagUpdate
+    }
+  }, [txBagUpdate])
+
+  useEffect(() => {
+    if (sigBagUpdate) {
+      setModalVisible(true)
+      setIsTransaction(false)
+      prevSigBag.current = sigBagUpdate
+    }
+  }, [sigBagUpdate])
+
+  return (
+    <React.Fragment>
+      {isTransaction
+        ? transactionBag && (
+            <SignTransactionFlow
+              apps={apps}
+              onClose={() => setModalVisible(false)}
+              transactionBag={transactionBag}
+              visible={modalVisible}
+              web3={web3}
+            />
+          )
+        : signatureBag && (
+            <SignMessageFlow
+              apps={apps}
+              onClose={() => setModalVisible(false)}
+              signatureBag={signatureBag}
+              visible={modalVisible}
+            />
+          )}
+    </React.Fragment>
```

Comment:
This condition is awkward... I do feel that separating these two "flows" is a good idea though, transactions are significantly more complicated than message signing, it doesn't make sense to smash their state together in one place as that only makes it more complicated.